### ISSUE TITLE
Swap! on Cursor should behave like Swap! on atom plus path

### ISSUE THREAD

**Issue by marcloyal3** [reagent-project/reagent#117]
```clojure
(def atm (atom {:kalle {:lelle 42}}))

(defn set-get
  ([k] (get-in @atm k))
  ([k v]
    (swap! atm assoc-in k v)
    (cond
      (= k [:kalle :lelle]) (println ":lelle was changed, cursor working!" @atm)
      )
    ))

(add-watch atm :a-key (fn [key ref old-val new-val] (println "atom working!!" @atm))) ;OK
(def cur (reagent/cursor set-get [:kalle :lelle]))
(swap! cur (fn [] (+ @cur 1)))
```

It seems that I should not have to dereference the cursor inside the anonymous function to get its value. It should be passed as the first argument to the anonymous function. For example, when using swap! with an atom I can simply do: (swap! atm update-in [:some-key :some-key-with-int-value] (fn [v] (+ v 1))) and v would be the value of the atom at the given path. Why couldn't I get the value of the atom at the cursor's path as an argument to the anonymous function when swap!-ing on cursors?

**Comment by JulianLeviston** [comment]
I don't really understand your issues here - perhaps you could elucidate further?

If you write your anonymous function so that it takes an argument instead of the zero-arity version you have here, it will do exactly what you want, won't it?

in other words, (swap! a (fn [prev] (assoc prev :something "yay"))) is identical to (swap! a assoc :something "yay").

**Comment by JulianLeviston** [comment]
I don't really understand your issues here - perhaps you could elucidate further?

If you write your anonymous function so that it takes an argument instead of the zero-arity version you have here, it will do exactly what you want, won't it?

in other words, (swap! a (fn [prev] (assoc prev :something "yay"))) is identical to (swap! a assoc :something "yay").

**Comment by marcloyal3** [comment]
I don't know how to make it more clear, but here is an attempt:

```clojure
(def atm (atom {:kalle {:lelle 42}}))

(defn set-get
  ([k] (get-in @atm k))
  ([k v]
    (swap! atm assoc-in k v)
    (cond
      (= k [:kalle :lelle]) (println ":lelle was changed, cursor working!" @atm)
      )
    ))

(add-watch atm :a-key (fn [key ref old-val new-val] (println "atom working!!" @atm))) ;OK
(def cur (reagent/cursor set-get [:kalle :lelle]))
(swap! cur (fn [arg] (println (str "arg passed in from swap! on cursor is: " a))))
(swap! atm (fn [arg] (println (str "arg passed in from swap! on atom is: " a))))
```

This will print:
arg passed in from swap! on cursor is:   (NOTHING!)
arg passed in from swap! on atom is: {:kalle {:lelle nil}}

Do you see what I mean? Why is swap! on cursor not passing in arg?

**Comment by JulianLeviston** [comment]
Um... actually it's working fine. I tried your code... and this is what it prints, which is what is should print:

    [Log] arg passed in from swap! on cursor is: 42 (core.js, line 52)
    [Log] atom working!! {:kalle {:lelle nil}} (core.js, line 52)
    [Log] :lelle was changed, cursor working! {:kalle {:lelle nil}} (core.js, line 52)
    [Log] arg passed in from swap! on atom is: {:kalle {:lelle nil}} (core.js, line 52)
    [Log] atom working!! nil (core.js, line 52)
    [Log] arg passed in from swap! on cursor is: 42 (core.js, line 52)
    [Log] atom working!! {:kalle {:lelle nil}} (core.js, line 52)
    [Log] :lelle was changed, cursor working! {:kalle {:lelle nil}} (core.js, line 52)
    [Log] arg passed in from swap! on atom is: {:kalle {:lelle nil}} (core.js, line 52)
    [Log] atom working!! nil (core.js, line 52)

If you're wondering why the nils, it's because you have side-effecting print statements as your return values for your swap functions (println returns nil), and swap assumes the function is pure (ie side-effect free), so it may run it many times.

Could you explain why you think that the cursor isn't passing arg in?

**Comment by JulianLeviston** [comment]
By the way, if you make the following small change (ie putting arg as your swap functions' return values):
  
    (swap! cur (fn [arg] (println (str "arg passed in from swap! on cursor is: " arg)) arg))
    (swap! atm (fn [arg] (println (str "arg passed in from swap! on atom is: " arg)) arg))
    
Then this is the result, which might perhaps be more in line with what you're expecting to see:

    [Log] arg passed in from swap! on cursor is: 42 (core.js, line 52)
    [Log] atom working!! {:kalle {:lelle 42}} (core.js, line 52)
    [Log] :lelle was changed, cursor working! {:kalle {:lelle 42}} (core.js, line 52)
    [Log] arg passed in from swap! on atom is: {:kalle {:lelle 42}} (core.js, line 52)
    [Log] atom working!! {:kalle {:lelle 42}} (core.js, line 52)
    [Log] arg passed in from swap! on cursor is: 42 (core.js, line 52)
    [Log] atom working!! {:kalle {:lelle 42}} (core.js, line 52)
    [Log] :lelle was changed, cursor working! {:kalle {:lelle 42}} (core.js, line 52)
    [Log] arg passed in from swap! on atom is: {:kalle {:lelle 42}} (core.js, line 52)
    [Log] atom working!! {:kalle {:lelle 42}} (core.js, line 52)

**Comment by marcloyal3** [comment]
Um... This doesn't address why it works in the case of the atom but doesn't work for cursor unless you do what you did. There are many ways to "make" it work but what I raised the issue for is that a swap! On cursor behaves differently from swap! On atom. You first reply did not mention that you made that small change. The issue I'm raising and trying to understand the real reason behind is why swap! Behavior is different between atom and cursor... when doing it the way I did without your little change. To me passing a cursor or an atom to swap! Should have identical behavior. So then without your change can you explain why the atom case prints the arg while the cursor case doesn't?

**Comment by antishok** [comment]
@marcloyal3 are you using reagent 0.5.0? or 0.5.0-alpha3?
I don't see the issue with 0.5.0, and it seems to have been dealt with in this [commit](https://github.com/reagent-project/reagent/commit/526dfa61ea87b80cb292dd96ca868303a000c41f) (which was merged *after* 0.5.0-alpha3 release)

**Comment by marcloyal3** [comment]
@antishok

Oh! thank you. I am new to Clojure/Script but not new to smelling bugs! So this helps a lot. 

In the latest iteration on alpha3 which is what I have been using, running the code below 

```clojure
(def atm (atom {:kalle {:lelle 42}}))

(defn set-get
  ([k] (get-in @atm k))
  ([k v]
    (swap! atm assoc-in k v)
    (cond
      (= k [:kalle :lelle]) (println ":lelle was changed, cursor working!" @atm)
      )
    ))

(add-watch atm :a-key (fn [key ref old-val new-val] (println "atom working!!" @atm))) ;OK
(def cur (reagent/cursor set-get [:kalle :lelle]))
(swap! atm (fn [arg] (println "arg passed in from swap! on atom is: " arg) arg))
(swap! cur (fn [arg] (println "arg passed in from swap! on cursor is: " arg) arg))

```

produces this (some lines removed):

arg passed in from swap! on atom is:  {:kalle {:lelle 42}}
arg passed in from swap! on cursor is:  nil

So I was really confused as to the difference in behavior....

Thanks a lot! 

Closing as its' been resolved in 5.0. I thought @homlsand said that 5.0 only had documentation changes. I'll go update all our deployed code from alpha3 to 5.0 :) 

:+1:

**Comment by JulianLeviston** [comment]
Cool :) Glad you got it resolved @marcloyal3 and thanks @antishok
